packages/react-native-editor/src/test/index.test.js

```diff
@@ -0,0 +1,185 @@
+
+/**
+ * WordPress dependencies
+ */
+import * as wpHooks from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import { registerGutenberg } from '..';
+import setupLocale from '../setup-locale';
+
+jest.mock( 'react-native/Libraries/ReactNative/AppRegistry' );
+jest.mock( '../setup-locale' );
+
+const initGutenberg = ( registerParams ) => {
+	let EditorComponent;
+	AppRegistry.registerComponent.mockImplementation(
+		( name, componentProvider ) => {
+			EditorComponent = componentProvider();
+		}
+	);
+	registerGutenberg( registerParams );
+
+	return render( <EditorComponent /> );
+};
+
+describe( 'Register Gutenberg', () => {
+	beforeEach( () => {
+		// We need to reset modules to guarantee that setup module is imported on every test.
+		jest.resetModules();
+	} );
+
+	it( 'registers Gutenberg editor component', () => {
+		registerGutenberg();
+		expect( AppRegistry.registerComponent ).toHaveBeenCalled();
+	} );
+
+	it( 'sets up locale before editor is initialized', () => {
+		const onModuleImported = jest.fn();
+		jest.doMock( '../setup', () => {
+			// To determine if the setup module is imported, we create a mock function that is called when the module is mocked.
+			onModuleImported();
+
+			return {
+				__esModule: true,
+				default: jest.fn().mockReturnValue( <></> ),
+			};
+		} );
+
+		initGutenberg();
+
+		// "invocationCallOrder" can be used to compare call orders between different mocks.
+		// Reference: https://git.io/JyBk0
+		const setupLocaleCallOrder = setupLocale.mock.invocationCallOrder[ 0 ];
+		const onSetupImportedCallOrder =
+			onModuleImported.mock.invocationCallOrder[ 0 ];
+
+		expect( setupLocaleCallOrder ).toBeLessThan( onSetupImportedCallOrder );
+	} );
+
+	it( 'beforeInit callback is invoked before the editor is initialized', () => {
+		const beforeInitCallback = jest.fn();
+		const onModuleImported = jest.fn();
+		jest.doMock( '../setup', () => {
+			// To determine if the setup module is imported, we create a mock function that is called when the module is mocked.
+			onModuleImported();
+
+			return {
+				__esModule: true,
+				default: jest.fn().mockReturnValue( <></> ),
+			};
+		} );
+
+		initGutenberg( { beforeInitCallback } );
+
+		// "invocationCallOrder" can be used to compare call orders between different mocks.
+		// Reference: https://git.io/JyBk0
+		const beforeInitCallOrder =
+			beforeInitCallback.mock.invocationCallOrder[ 0 ];
+		const onSetupImportedCallOrder =
+			onModuleImported.mock.invocationCallOrder[ 0 ];
+
+		expect( beforeInitCallOrder ).toBeLessThan( onSetupImportedCallOrder );
+	} );
+
+	it( 'dispatches "native.pre-render" hook before the editor is rendered', () => {
+		const doAction = jest.spyOn( wpHooks, 'doAction' );
+
+		// An empty component is provided in order to listen for render calls of the editor component.
+		const onRenderEditor = jest.fn();
+		const EditorComponent = () => {
+			onRenderEditor();
+			return null;
+		};
+		jest.mock( '../setup', () => ( {
```

Comment:
Each of the mocks of `../setup` before this point rely upon `doMock`. Is the usage of `mock` here and within the rest of the tests an oversight? If not, what is the reason to diverge in usage?

Reply:
The reason for using `doMock` over `mock` is related to passing out-of-scope variables, like the case of `onModuleImported`:
https://github.com/WordPress/gutenberg/blob/61b800d3ef47a5637848174b7f5cd594350b9bf1/packages/react-native-editor/src/test/index.test.js#L46-L49

Otherwise, using `mock` produces the following error:
```
ReferenceError: /gutenberg-mobile/gutenberg/packages/react-native-editor/src/test/index.test.js: The module factory of `jest.mock()` is not allowed to reference any out-of-scope variables.
```

Using `doMock` avoids this error but I've just realized after checking [the documentation](https://jestjs.io/docs/es6-class-mocks#calling-jestmock-with-the-module-factory-parameter) that there's another option by using `mock` and prefixing the word `mock` to the variable.

_Extracted from Jest documentation:_

> A limitation with the factory parameter is that, since calls to jest.mock() are hoisted to the top of the file, it's not possible to first define a variable and then use it in the factory. An exception is made for variables that start with the word 'mock'. It's up to you to guarantee that they will be initialized on time!

Reply:
@dcalhoun I changed `doMock` by the `mock` function in https://github.com/WordPress/gutenberg/pull/37073/commits/74b1756c89371f48e581f906d46c055fa46ca177, thanks for raising this question as it helped me to find an alternative I didn't spot at the beginning 😅 .